experimental/devices/bmp180/bmp180.go

```diff
@@ -0,0 +1,259 @@
+	cmdStartPressureConv uint8 = (1 << ctrlStartConversionShift) | (0x14 << ctrlMeasurementControlShift)
+)
+
+const (
+	tempConvTime = 4500 * time.Microsecond // maximum conversion time for temperature
+)
+
+// maximum conversion time for pressure
+var pressureConvTime = [...]time.Duration{
+	4500 * time.Microsecond,
+	7500 * time.Microsecond,
+	13500 * time.Microsecond,
+	25500 * time.Microsecond,
+}
+
+// Sense returns measurements as °C and kPa.
+func (d *Dev) Sense(env *devices.Environment) error {
+	// start conversion for temperature
+	if err := d.dev.WriteUint8(regCtrlMeas, cmdStartTempConv); err != nil {
+		return err
+	}
+
+	time.Sleep(tempConvTime)
+
+	// read value
+	ut, err := d.dev.ReadUint16(regOutMSB)
+	if err != nil {
+		return err
+	}
+
+	temp := d.cal.compensateTemp(ut)
+
+	// start conversion for pressure
+	cmd := cmdStartPressureConv
+	cmd |= uint8(d.os) << ctrlOversamplingShift
+
+	if err := d.dev.WriteUint8(regCtrlMeas, cmd); err != nil {
+		return err
+	}
+
+	time.Sleep(pressureConvTime[d.os])
+
+	// read value
+	var pressureBuf [3]byte
+	if err := d.dev.ReadStruct(regOutMSB, pressureBuf[:]); err != nil {
+		return err
+	}
+
+	up := (int32(pressureBuf[0])<<16 + int32(pressureBuf[1])<<8 | int32(pressureBuf[2])) >> (8 - d.os)
+
+	pressure := d.cal.compensatePressure(up, int32(ut), d.os)
+
+	env.Temperature = devices.Celsius(temp * 100)
+	env.Pressure = devices.KPascal(pressure)
+	return nil
+}
+
+// Halt is a noop for the BMP180.
+func (d *Dev) Halt() error {
+	return nil
+}
+
+const (
+	chipAddress = 0x77 // the address of a BMP180 is fixed at 0x77
+	chipID      = 0x55 // contents of the ID register, always 0x55
+)
+
+// registers
+const (
+	regChipID           = 0xD0 // register contains the chip id
+	regCalibrationStart = 0xAA // first calibration register address
+	regSoftReset        = 0xE0 // soft reset register
+	regCtrlMeas         = 0xF4 // control measurement
+	regOutMSB           = 0xF6 // MSB of measurement
+	regOutLSB           = 0xF7 // LSB of measurement
+	regOutXLSB          = 0xF8 // extended LSB of measurement
+)
+
+const softResetValue = 0xB6
+
+// New returns an object that communicates over I²C to BMP180 environmental
+// sensor. The frequency for the bus can be up to 3.4MHz.
+func New(b i2c.Bus, os Oversampling) (d *Dev, err error) {
+	bus := &i2c.Dev{Bus: b, Addr: chipAddress}
+	d = &Dev{
+		os: os,
+		dev: &mmr.Dev8{
+			Conn:  bus,
+			Order: binary.BigEndian,
+		},
+	}
+
+	id, err := d.dev.ReadUint8(regChipID)
+	if err != nil {
+		return nil, err
+	}
+
+	if id != chipID {
+		return nil, fmt.Errorf("bmp180: unexpected chip id 0x%x; is this a BMP180?", id)
+	}
+
+	// issue soft reset to initialize device
+	if err := d.dev.WriteUint8(regSoftReset, softResetValue); err != nil {
+		return nil, err
+	}
+
+	// wait for restart
+	time.Sleep(10 * time.Millisecond)
+
+	// read calibration data from internal EEPROM, 11 registers with two bytes each
+	var cal [22]byte
```

Comment:
The main reason I prefer this is because of the current code use reflection twice. It is mostly a bug in the API design of `mmr` but still. The fact that big endian decoding is done part manually, part automatically, is kind of weird.

```
func isValid(i int16) bool {
  return i != 0 && i != ^int16(0)
}
func isValidU(i uint16) bool {
  return i != 0 && i != 0xFFFF
}

func (c*calibration) isValid() bool {
  return isValid(c.AC1) && isValid(c.AC2) && isValid(c.AC3) && isValidU(c.AC4) && isValidU(c.AC5) && isValidU(c.AC6) && isValid(c.B1) && isValid(c.B2) && isValid(c.MB) && isValid(c.MC) && isValid(c.MD)
}
```